File: src/main/java/sensormodels/LightSensorData.java
package sensormodels;

import com.google.gson.annotations.Expose;
import com.google.gson.annotations.SerializedName;
import database.LuceneManager;
import java.io.File;
import java.sql.PreparedStatement;
import java.sql.SQLException;
import java.util.Date;
import org.apache.lucene.document.Document;
import org.apache.lucene.document.Field;
import org.apache.lucene.document.StringField;
import org.apache.lucene.document.TextField;
import org.bson.codecs.pojo.annotations.BsonIgnore;
import utils.WebAppConstants;

public class LightSensorData extends DatabaseModel {

  @BsonIgnore public static final String MY_SQL_TABLE_NAME = "LightSensorData";

  @BsonIgnore public static final String FILE_NAME = "LightSensor";

  @SerializedName("sensor_name")
  @Expose
  private String sensorName;

  @SerializedName("timestamp")
  @Expose
  private String timestamp;

  @SerializedName("sensor_data")
  @Expose
  private SensorData sensorData;

  private String luxValue;

  @SerializedName("formatted_date")
  @Expose
  private String formattedDate;

  @BsonIgnore private File file;

  public String getLuxValue() {
    return luxValue;
  }

  public void setLuxValue(String luxValue) {
    this.luxValue = luxValue;
  }

  public String getSensorName() {
    return sensorName;
  }

  public void setSensorName(String sensorName) {
    this.sensorName = sensorName;
  }

  public String getTimestamp() {
    return timestamp;
  }

  public void setTimestamp(String timestamp) {
    this.timestamp = timestamp;
  }

  public SensorData getSensorData() {
    return sensorData;
  }

  public void setSensorData(SensorData sensorData) {
    this.sensorData = sensorData;
  }

  public void setFormattedDate() {
    this.formattedDate = WebAppConstants.inputDateFormat.format(new Date(timestamp));
  }

  @Override
  public String getStartTime() {
    return this.getTimestamp();
  }

  public String getFormattedDate() {
    return formattedDate;
  }

  @Override
  @BsonIgnore
  public Document getDocument() {
    Document doc = new Document();
    if (this.getLuxValue() != null) {
      doc.add(
          new TextField(LuceneManager.LuceneConstants.LUX, this.getLuxValue(), Field.Store.YES));
    }
    doc.add(
        new StringField(
            LuceneManager.LuceneConstants.SENSOR_NAME, this.getSensorName(), Field.Store.YES));
    doc.add(
        new StringField(
            LuceneManager.LuceneConstants.FORMATTED_DATE,
            this.getFormattedDate(),
            Field.Store.YES));
    //         use a string field for timestamp because we don't want it tokenized
    doc.add(
        new StringField(
            LuceneManager.LuceneConstants.TIMESTAMP, this.getTimestamp(), Field.Store.YES));
    return doc;
  }

  @Override
  @BsonIgnore
  public String getMongoCollectionName() {
    return "LightSensorData";
  }

  @Override
  @BsonIgnore
  public Class<LightSensorData> getClassObject() {
    return LightSensorData.class;
  }

  @Override
  @BsonIgnore
  public String getTableName() {
    return MY_SQL_TABLE_NAME;
  }

  @Override
  @BsonIgnore
  public String getCreateTableQuery() {
    return "CREATE TABLE "
        + this.getTableName()
        + "(timestamp VARCHAR(30) , "
        + " formatted_date VARCHAR(10) , "
        + " sensor_name VARCHAR(30) , "
        + " lux INTEGER) ";
  }

  @Override
  @BsonIgnore
  public String getInsertIntoTableQuery() {
    return " insert into "
        + this.getTableName()
        + " (timestamp, formatted_date, sensor_name,lux)"
        + " values (?, ?, ?, ?)";
  }

  @Override
  @BsonIgnore
  public void fillQueryData(PreparedStatement preparedStmt) throws SQLException {
    preparedStmt.setString(1, this.getTimestamp());
    preparedStmt.setString(2, this.getFormattedDate());
    preparedStmt.setString(3, this.getSensorName());
    preparedStmt.setInt(4, this.getSensorData().getLux());
  }

  @Override
  @BsonIgnore
  public String getFileName() {
    return FILE_NAME;
  }

  @Override
  public void setFile(File file) {
    this.file = file;
  }

  @Override
  @BsonIgnore
  public File getFile() {
    return this.file;
  }

  public static class SensorData {
    @SerializedName("lux")
    @Expose
    private Integer lux;

    public Integer getLux() {
      return lux;
    }

    public void setLux(Integer lux) {
      this.lux = lux;
    }
  }
}

File: src/main/java/sensormodels/LightSensorData.kt
package sensormodels

import com.google.gson.annotations.Expose
import com.google.gson.annotations.SerializedName
import database.LuceneManager
import org.apache.lucene.document.Document
import org.apache.lucene.document.Field
import org.apache.lucene.document.StringField
import org.apache.lucene.document.TextField
import org.bson.codecs.pojo.annotations.BsonIgnore
import sensormodels.store.models.SuperStoreModel
import utils.WebAppConstants
import java.io.File
import java.sql.PreparedStatement
import java.util.*

data class LightSensorData(
    @SerializedName("sensor_name")
    @Expose
    var sensorName: String? = null,
    @SerializedName("timestamp")
    @Expose
    var timestamp: String? = null,
    @SerializedName("sensor_data")
    @Expose
    var sensorData: SensorData? = null,
    var luxValue: String? = null
) : SuperStoreModel {
    @SerializedName("formatted_date")
    @Expose
    var formattedDate: String? = null
        private set
    override val fileName = FILE_NAME

    @BsonIgnore
    override var file: File? = null

    override fun setFormattedDate() {
        formattedDate = WebAppConstants.inputDateFormat.format(Date(timestamp))
    }

    override val startTime: String? = timestamp

    data class SensorData(
        @SerializedName("lux")
        @Expose
        var lux: Int? = null
    )

    companion object {
        @BsonIgnore
        val MY_SQL_TABLE_NAME = "LightSensorData"

        @BsonIgnore
        val FILE_NAME = "LightSensor"
    }

    override val document = Document().apply {
        if (luxValue != null) {
            add(
                TextField(LuceneManager.LuceneConstants.LUX, luxValue, Field.Store.YES)
            )
        }
        add(
            StringField(
                LuceneManager.LuceneConstants.SENSOR_NAME, sensorName, Field.Store.YES
            )
        )
        add(
            StringField(
                LuceneManager.LuceneConstants.FORMATTED_DATE,
                formattedDate,
                Field.Store.YES
            )
        )
        //         use a string field for timestamp because we don't want it tokenized
        add(
            StringField(
                LuceneManager.LuceneConstants.TIMESTAMP, timestamp, Field.Store.YES
            )
        )
    }
    override val mongoCollectionName = "LightSensorData"
    override val tableName = MY_SQL_TABLE_NAME
    override val createTableQuery =
        "CREATE TABLE ${this.tableName}(timestamp VARCHAR(30) ,  formatted_date VARCHAR(10) ,  sensor_name VARCHAR(30) ,  lux INTEGER) "
    override val insertIntoTableQuery =
        "INSERT into ${this.tableName} (timestamp, formatted_date, sensor_name,lux) values (?, ?, ?, ?)"

    override fun fillQueryData(preparedStmt: PreparedStatement) {
        preparedStmt.apply {
            setString(1, timestamp)
            setString(2, formattedDate)
            setString(3, sensorName)
            setInt(4, sensorData!!.lux!!)
        }
    }
}